### PR TITLE
add -g option for openssl pcre zlib

### PR DESCRIPTION
**Pull request by zhuizhuhaomeng**
I hereby granted the copyright of the changes in this pull request
to the authors of this openresty project.

auto add  -g  option no matter  user config with or without --with-(openssl|zlib|pcre)-opt

### CODE REVIEW

File: util/configure [hunk 1]
```diff
@@ -359,6 +371,15 @@ for my $opt (@ARGV) {
     }
 }
 
+
+if ($OS eq "linux") {
+        for my $lib (qw(openssl pcre zlib)) {
```
Comment (doujiang24) [inline util/configure:376]: style: bad indent.

File: util/configure [hunk 2]
```diff
@@ -359,6 +369,12 @@ for my $opt (@ARGV) {
     }
 }
 
+for my $lib (qw(openssl pcre zlib)) {
```
Comment (agentzh) [inline util/configure:372]: Better use `qw/.../` instead of `qw(...)` for better readability here.

File: util/configure [hunk 3]
```diff
@@ -359,6 +369,12 @@ for my $opt (@ARGV) {
...
+    if (exists($ngx_libs{$lib}) && not exists($ngx_libs_opt{$lib."-opt"})) {
```
Comment (agentzh) [inline util/configure:373]: Please use `!` instead of `not` for expression level operators.

File: util/configure [hunk 4]
```diff
@@ -359,6 +369,12 @@ for my $opt (@ARGV) {
...
+        push @ngx_opts, "--with-".$lib."-opt=-g";
```
Comment (agentzh) [inline util/configure:374]: Please always use spaces around binary operators like `.`.

File: util/configure [hunk 5]
```diff
@@ -346,6 +349,13 @@ for my $opt (@ARGV) {
         }
         $path = File::Spec->rel2abs($path);
         push @ngx_opts, "--with-$lib=$path";
+        $ngx_libs{$lib} = "y";
+
+    } elsif ($opt =~ /^--with-(openssl|pcre|zlib)-opt=(.*)/) {
+        my ($lib, $opt) = ($1, $2);
+        $opt.=" -g";
```
Comment (agentzh) [inline util/configure:356]: Missing spaces around binary operator `.=`. Please fix other similar issues in your patch here. I see several such things below as well.

File: util/configure (hunk 4 above)
Comment (agentzh) [reply util/configure:374]: The logic here does not look quite right. Even if the user specifies her own `--with-xxx-opt`, we should still prepend `-g` to the option string.

File: util/configure [hunk 6]
```diff
@@ -346,6 +349,13 @@ for my $opt (@ARGV) {
...
+        push @ngx_opts, "--with-$lib-opt=$opt";
+        $ngx_libs_opt{$lib."-opt"} = "y";
```
Comment (agentzh) [inline util/configure:358]: We are not using the value `y` in any way, so a Perl boolean value `1` should be enough here?

File: util/configure [hunk 7]
```diff
@@ -346,6 +349,12 @@ for my $opt (@ARGV) {
         }
         $path = File::Spec->rel2abs($path);
         push @ngx_opts, "--with-$lib=$path";
+        $with_ext_lib{$lib} = 1;
+
+    } elsif ($opt =~ /^--with-(openssl|pcre|zlib)-opt=(.*)/) {
+        my ($lib, $opt) = ($1, $2);
+        push @ngx_opts, "--with-$lib-opt=-g $opt";
+        $with_ext_lib_opts{"$lib-opt"} = 1;
```
Comment (agentzh) [inline util/configure:357]: We don't need to add the `-opt` suffix to the hash table key. When we look up this table, we use the key of `$lib` anyway. It's wasteful to always append this artificial `-opt` suffix for no gains.

File: util/configure [hunk 8]
```diff
@@ -359,6 +368,12 @@ for my $opt (@ARGV) {
     }
 }
 
+for my $lib (qw/openssl pcre zlib/) {
+    if ($with_ext_lib{$lib} && !$with_ext_lib_opts{"$lib-opt"}) {
```
Comment (agentzh) [inline util/configure:372]: Ditto, let's save the trouble of constructing the key `"$lib-opt"`. Just use `$lib` instead.

**Comment by agentzh** [comment]
Merged. Thanks.

**Pull request closed without merging**